### PR TITLE
gitserver: support max retries for sg maintenance

### PR DESCRIPTION
**Pull request by stefanhengl**
So far, sg maintenance followed the same logic as git gc. If a log
file (sgm.log or gc.log) is present and the repo is old enough, we
reclone immediately.

However, recloning is expensive for monorepos and we want to avoid it as
much as possible. Hence we switch sg maintenance to a different
strategy.

Customers can now set the ENV var `SRC_SGM_RETRIES` which defines the
number of times we retry sg maintenance before recloning. The default is
"-1" which means "infinite". 

This removes the tight coupling between repoTTLSGM and sgmLogExpire, which
makes the process easier to reason about.

git gc remains unchanged to allow customers to fall back to the legacy setup.

Relates to https://github.com/sourcegraph/sourcegraph/pull/33432

## Test plan
- new unit test
- manual testing:
  - Ran a local instance of Sourcegraph
  - corrupted a `<repo>` with `touch ~/.sourcegraph/repos/github.com/<repo>/.git/refs/heads/deadbeaf`
  - confirmed that sgm.log is created
  - confirmed that the counter increments
  - corrupted the sgm.log file and confirmed that the next cleanup run fixes the file with a count at 1
  - confirmed that the repo is recloned if `failed > SRC_SGM_RETRIES`
  - confirmed that sgm.log is gone and is not recreated

### CODE REVIEW

File: cmd/gitserver/server/cleanup.go [hunk 1]
```diff
@@ -877,6 +878,46 @@ func gitGC(dir GitDir) error {
...
+	}
+
+	b = b[from+len(prefix):]
+	if to := bytes.IndexByte(b, '\n'); to > 0 {
+		b = b[:to]
+	}
+
+	n, _ := strconv.Atoi(string(b))
+	return n
+}
```
Comment (keegancsmith) [inline cmd/gitserver/server/cleanup.go:919]: this looks familiar :)

File: cmd/gitserver/server/cleanup.go [hunk 2]
```diff
@@ -877,6 +878,46 @@ func gitGC(dir GitDir) error {
...
+// writeSGMLog writes a log file with the format
+// 		<sgmLogPrefix>=<int>
+//
+// 		<error message>
+//
+func writeSGMLog(dir GitDir, m []byte) error {
+	return os.WriteFile(
+		dir.Path(sgmLog),
+		[]byte(fmt.Sprintf("%s%d\n\n%s\n", sgmLogPrefix, bestEffortReadFailed(dir)+1, m)),
+		0666,
```
Comment (keegancsmith) [inline cmd/gitserver/server/cleanup.go:892]: what masks do we use in other places? Do we want the other group to read this? ie should this be `0660`?

File: cmd/gitserver/server/cleanup.go [hunk 3]
```diff
@@ -877,6 +878,46 @@ func gitGC(dir GitDir) error {
...
+
+// writeSGMLog writes a log file with the format
+// 		<sgmLogPrefix>=<int>
+//
+// 		<error message>
+//
+func writeSGMLog(dir GitDir, m []byte) error {
+	return os.WriteFile(
+		dir.Path(sgmLog),
+		[]byte(fmt.Sprintf("%s%d\n\n%s\n", sgmLogPrefix, bestEffortReadFailed(dir)+1, m)),
```
Comment (keegancsmith) [inline cmd/gitserver/server/cleanup.go:891]: should we include a human readable description of what the file is for curious admins that run `cat`

File: cmd/gitserver/server/cleanup_test.go [hunk 4]
```diff
@@ -1317,3 +1321,51 @@ func TestSGMLogFile(t *testing.T) {
...
+		{
+			content: []byte(`failed=
+
+error message`),
+			wantRetries: 0,
+		},
+		{
+			content: []byte(`failed=deadbeaf
+
+error message`),
```
Comment (keegancsmith) [inline cmd/gitserver/server/cleanup_test.go:1359]: add a test case with failed= twice

**Review by keegancsmith — APPROVED** [review]
Can you detail what your manual tests where in your test plan? This helps others in future when making changes here.

**Review by ryanslade — APPROVED** [review]

File: cmd/gitserver/server/cleanup.go (hunk 1 above)
Comment (stefanhengl) [reply cmd/gitserver/server/cleanup.go:919]: :) copying from the best

File: cmd/gitserver/server/cleanup.go (hunk 2 above)
Comment (stefanhengl) [reply cmd/gitserver/server/cleanup.go:892]: We could even go with `0600` like we do for `sg_refhash`.

File: cmd/gitserver/server/cleanup.go (hunk 3 above)
Comment (stefanhengl) [reply cmd/gitserver/server/cleanup.go:891]: Yes, let's be nice. I will add something.

**Pull request merged**